packages/server/src/complete/candidates/createTableCandidates.ts

```diff
@@ -10,20 +10,54 @@ import { ICONS } from '../CompletionItemUtils'
  * @returns
  */
 function allTableNameCombinations(table: Table): string[] {
-  const names = [table.tableName]
-  if (table.database) names.push(table.database + '.' + table.tableName)
-  if (table.catalog)
-    names.push(table.catalog + '.' + table.database + '.' + table.tableName)
-  return names
+  if (table.catalog) {
+    return [table.catalog + '.' + table.database + '.' + table.tableName]
+  }
+  if (table.database) {
+    return [table.database + '.' + table.tableName]
+  }
+  return [table.tableName]
 }
 
 export function createTableCandidates(
   tables: Table[],
   lastToken: string,
   onFromClause?: boolean
 ) {
-  return tables
-    .flatMap((table) => allTableNameCombinations(table))
+  const qualificationLevel = lastToken.split('.').length - 1
+
+  const qualifiedEntities = tables.flatMap((table) => {
+    let qualificationNeeded = 0
+    if (table.catalog) {
+      qualificationNeeded++
+    }
+    if (table.database) {
+      qualificationNeeded++
+    }
+    const qualificationLevelNeeded = qualificationNeeded - qualificationLevel
+    switch (qualificationLevelNeeded) {
+      case 0:
+        return allTableNameCombinations(table)
+      case 1:
+        if (table.catalog && table.database) {
+          return [table.catalog + '.' + table.database]
+        }
+        if (table.database) {
+          return [table.database]
+        }
+        break
+      case 2:
+        if (table.catalog) {
+          return [table.catalog]
+        }
+        break
+    }
+    return []
+  })
+
+  const uniqueEntities = [...new Set(qualifiedEntities)]
```

Comment:
question: Just trying to wrap my head around this. Given that we are creating a set out of `qualifiedEntities` I assume that we could end up with duplicates. Could you give an example where that could happen please?

Reply:
ehhhh... I'm not sure actually :D I rewrote the code here a few times so it's possible I left this here by accident. It looks like it shouldn't be needed so I'll delete it